GameEngine/LinqExtensions.cs

```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace GameEngine
+{
+    public static class LinqExtensions
+    {
+        public static T RemoveMin<T>(this ICollection<T> collection, Func<T, int> valueFunction)
```

Comment:
💡 An alternative to this extension method would be to have SearchNode implement IComparable. Thoughts?

Something like:
```csharp
        public int CompareTo(object obj)
        {
            if (obj == null) return 1;

            SearchNode other = obj as SearchNode;
            if (other != null)
                return H().CompareTo(other.H());
            else
                throw new ArgumentException("Object is not a SearchNode");
        }
```

Reply:
Comparison is weird in a vacuum.  Each search algorithm is going to want to induce a different ordering over the game states.  You can get there as you suggest, but you'd need to specialize search nodes for each agent and write IComparable for each one of those specializations. I think that would be more C#-y, but I think this reads clearly enough as it is too.

Reply:
Good point. I'm fine with it either way.

Reply:
I considered going the `IComparable` route, but like Bass pointed out, different agents will want to order their frontier differently. We should probably use something more domain-specific to encapsulate choosing the next state in a clearer way.